### PR TITLE
Pull request for updates/review

### PR DESCRIPTION
**Pull request by silkycode**

### CODE REVIEW

File: src/main/java/StringManipulation.java [hunk 1]
```diff
@@ -1,33 +1,116 @@
...
+
     @Override
     public String getString() {
-        return null;
+        return myString;
     }
 
     @Override
     public void setString(String string) {
+        myString = string;
```
Comment (Hester1937) [inline src/main/java/StringManipulation.java:20]: this.myString as it is a property of the class.

File: src/main/java/StringManipulation.java [hunk 2]
```diff
@@ -1,33 +1,116 @@
...
     }
 
     @Override
     public int count() {
-        return 0;
+        if (myString == null || myString.length() == 0) {
+            return 0;
```
Comment (Hester1937) [inline src/main/java/StringManipulation.java:26]: Input Validation: The code correctly checks if the input string (myString) is null or empty. Returning 0 in these cases seems appropriate, assuming that an empty string should not be counted as a word.

File: src/main/java/StringManipulation.java [hunk 3]
```diff
@@ -1,33 +1,116 @@
...
+        }
+        int wordCount = 0;
+        // Use both whitespace and half thin-space characters to slice strings
+        String[] tokens = myString.split(" | ");
+        for (String i : tokens) {
+            if (Pattern.matches("^[A-Za-z]+$", i)) {
+                wordCount++;
+            }
+        }
+        return wordCount;
```
Comment (Hester1937) [inline src/main/java/StringManipulation.java:36]: The code iterates over the tokens and checks if each token consists only of alphabetical characters using a regular expression pattern. This pattern matches strings that contain only one or more uppercase or lowercase letters. This approach assumes that words are defined as sequences of alphabetic characters without any punctuation or numbers. If this definition aligns with your requirements, then the code is suitable.

File: src/main/java/StringManipulation.java [hunk 4]
```diff
@@ -1,33 +1,116 @@
...
     public String removeNthCharacter(int n, boolean maintainSpacing) {
-        return null;
+        if (n <= 0) {
+            throw new IllegalArgumentException("n is invalid: n > 0");
+        } else if (myString == null) {
+            throw new NullPointerException("String has not been initialized yet, or is null");
+        } else if (n > myString.length()) {
+            throw new IndexOutOfBoundsException("n cannot exceed length of parameter String");
+        }
+
```
Comment (Hester1937) [inline src/main/java/StringManipulation.java:48]: throws appropriate exceptions for invalid cases.

File: src/main/java/StringManipulation.java [hunk 5]
```diff
@@ -1,33 +1,116 @@
...
+        ArrayList<Character> characters = new ArrayList<Character>();
```
Comment (Hester1937) [inline src/main/java/StringManipulation.java:49]: Instead of new ArrayList<Character>(), you can use new ArrayList<>() to let the compiler infer the type.

File: src/main/java/StringManipulation.java [hunk 6]
```diff
@@ -1,33 +1,116 @@
...
+            throw new IllegalArgumentException("startWord and endWord cannot be less than 1");
+        } else if (endWord < startWord) {
+            throw new IndexOutOfBoundsException("index of endWord cannot be prior to index of startWord");
+        }
+
+        int numberOfWords = (endWord - startWord) + 1;
+        String[] words = myString.split(" | ");
+        String[] subWords = new String[numberOfWords];
+        for (int i = 0, j = startWord - 1; i < numberOfWords; i++, j++) {
+            subWords[i] = words[j];
```
Comment (Hester1937) [inline src/main/java/StringManipulation.java:84]: Instead of using a fixed-size array subWords, consider using a List<String> implementation, such as an ArrayList, to allow for dynamic resizing as needed. This eliminates the need to calculate the number of words beforehand.

File: src/main/java/StringManipulation.java [hunk 7]
```diff
@@ -1,33 +1,116 @@
...
+            // Check for duplicate integers
+            for (int j = i + 1; j < indices.length; j++) {
+                if (indices[i] == indices[j]) {
+                    throw new IllegalArgumentException("passed index array must contain unique integer values");
+                }
+            }
+        }
+        char[] charString = myString.toCharArray();
+        char[] referenceString = myString.toCharArray();
+        for (int i = 0; i < charString.length; i++) {
```
Comment (Hester1937) [inline src/main/java/StringManipulation.java:111]: The code creates two separate character arrays, charString and referenceString, from myString.toCharArray(). However, you can directly modify charString in place without the need for the referenceString array.

File: src/test/java/StringManipulationTest.java [hunk 8]
```diff
@@ -19,131 +24,209 @@ public void tearDown() {
...
+        int[] array = new int[]{1,2,7,4,5,3};
+        manipulatedstring.setString("string");
+        IndexOutOfBoundsException testException = assertThrows(IndexOutOfBoundsException.class, () -> {
+            manipulatedstring.restoreString(array);
+        });
+        assertEquals("Array of indices cannot contain negative integers or integers greater than the length of the current String", testException.getMessage());
     }
-
 }
+
```
Comment (Hester1937) [inline src/test/java/StringManipulationTest.java:232]: In the test class, add more test cases to cover different scenarios and edge cases.
such as:
assertEquals("bcdea", stringManipulator.restoreString(new int[]{1, 2, 3, 4, 0}));
assertEquals("", stringManipulator.restoreString(new int[]{}));
assertEquals("Hello, World!", stringManipulator.restoreString(new int[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13}));
Additional test cases help ensure the code handles different inputs correctly and increase the overall test coverage.

Comment (Hester1937) [reply src/test/java/StringManipulationTest.java:232]: Overall Recap: These improvements result in a more robust and efficient code implementation, with enhanced error handling and increased test coverage. It's important to thoroughly test the code and consider edge cases to ensure its correctness and reliability.

**Review by Hester1937 — APPROVED** [review]
